File: src/main/scala/org/dupontmanual/image/Stack.scala
package org.dupontmanual.image

import math.{ abs, min, max }
import scalafx.geometry.Rectangle2D
import scalafx.geometry.Bounds
import scalafx.geometry.BoundingBox
import scalafx.scene.image.ImageView
import scalafx.scene.layout.Pane
import scalafx.scene.{ Group, Node }
import scalafx.scene.shape.{ Rectangle => SfxRectangle, Shape }
import scalafx.scene.transform.{ Transform => SfxTransform }
import scalafx.application.Platform
import scalafx.concurrent.Task

/** represent two images, one in front of the other */
/* private[image] */ class Stack(front: Image, back: Image, xAlign: XAlign, yAlign: YAlign, dx: Double, dy: Double) extends Image {
  def this(front: Image, back: Image, align: Align, dx: Double, dy: Double) = {
    this(front, back, align.xAlign, align.yAlign, dx, dy)
  }

  val backBounds = Stack.translateRect(back.displayBounds, 
		  							   Stack.dx(back.displayBounds, xAlign, 0), 
		  							   Stack.dy(back.displayBounds, yAlign, 0))
  val frontBounds = Stack.translateRect(front.displayBounds,
             							Stack.dx(front.displayBounds, xAlign, dx),
             							Stack.dy(front.displayBounds, yAlign, dy))
  val newLeft = min(backBounds.minX, frontBounds.minX)
  val newRight = max(backBounds.minX + backBounds.width, frontBounds.minX + frontBounds.width)
  val newTop = min(backBounds.minY, frontBounds.minY)
  val newBottom = max(backBounds.maxY, frontBounds.maxY)
  val newBounds = new BoundingBox(newLeft, newTop, newRight - newLeft, newBottom - newTop)
  val backDx = backBounds.minX - newLeft
  val backDy = backBounds.minY - newTop
  val frontDx = frontBounds.minX - newLeft
  val frontDy = frontBounds.minY - newTop
  val backTransform = SfxTransform.translate(backDx, backDy)
  val frontTransform = SfxTransform.translate(frontDx, frontDy)
  
  def bounds(): Shape = SfxRectangle(newBounds.minX, newBounds.minY, newBounds.width, newBounds.height)

  def buildImage(): Node = {
    def newNode() = {
      val backView = new Group(back.buildImage()) { transforms = List(backTransform) }
      val frontView = new Group(front.buildImage()) { transforms = List(frontTransform) }
      new Pane {
        children = new Group(backView, frontView)
        prefWidth = newBounds.width
        prefHeight = newBounds.height
      }
    }
    if (Platform.isFxApplicationThread) {
      newNode()
    } else {
      val theNode = Task[Node] { newNode() }
      Platform.runLater(theNode)
      theNode.get()
    }    
  }
}

private[image] object Stack {
  def apply(front: Image, back: Image) = new Stack(front, back, Align.Center, 0.0, 0.0)
  def apply(front: Image, back: Image, alignment: Align) = 
    new Stack(front, back, alignment.xAlign, alignment.yAlign, 0.0, 0.0)
  def apply(front: Image, back: Image, xAlign: XAlign, yAlign: YAlign) =
    new Stack(front, back, xAlign, yAlign, 0.0, 0.0)
  def apply(front: Image, back: Image, dx: Double, dy: Double) =
    new Stack(front, back, XAlign.Center, YAlign.Center, dx, dy)
  def apply(front: Image, back: Image, xAlign: XAlign, yAlign: YAlign, dx: Double, dy: Double) =
    new Stack(front, back, xAlign, yAlign, dx, dy)
  def apply(front: Image, back: Image, alignment: Align, dx: Double, dy: Double) =
    new Stack(front, back, alignment.xAlign, alignment.yAlign, dx, dy)

  def translateRect(rect: BoundingBox, dx: Double, dy: Double): BoundingBox = {
    new BoundingBox(rect.minX + dx, rect.minY + dy, rect.width, rect.height)
  }

  def centerOf(bounds: BoundingBox): Point = {
    Point(0.5 * (bounds.minX + bounds.maxX), 0.5 * (bounds.minY + bounds.maxY))
  }
  
  def dx(bds: BoundingBox, xAlign: XAlign, xOffset: Double): Double = xOffset + (xAlign match {
    case XAlign.Left => 0
    case XAlign.Center => -0.5 * bds.width
    case XAlign.Right => -bds.maxX
  })
  
  def dy(bds: BoundingBox, yAlign: YAlign, yOffset: Double): Double = yOffset + (yAlign match {
    case YAlign.Top => 0
    case YAlign.Center => -0.5 * bds.height
    case YAlign.Bottom => -bds.maxY
  })
}

File: src/main/java/net/toddobryan/java/image/Stack.java
package net.toddobryan.java.image;

import javafx.application.Platform;
import javafx.concurrent.Task;
import javafx.geometry.BoundingBox;
import javafx.geometry.Bounds;
import javafx.scene.Group;
import javafx.scene.Node;
import javafx.scene.layout.Pane;
import javafx.scene.shape.Rectangle;
import javafx.scene.shape.Shape;
import javafx.scene.transform.Transform;

public class Stack extends Image {
	private Image front;
	private Image back;
	
	private Bounds newBounds;
	private Transform backTransform;
	private Transform frontTransform;
	
	public Stack(Image front, Image back, XAlign xAlign, YAlign yAlign, double dx, double dy) {
		this.front = front;
		this.back = back;
		
		BoundingBox backBounds = Stack.translateRect(
				back.getDisplayBounds(), 
				Stack.getDx(back.getDisplayBounds(), xAlign, 0),
				Stack.getDy(back.getDisplayBounds(), yAlign, 0));
		BoundingBox frontBounds = Stack.translateRect(
				front.getDisplayBounds(), 
				Stack.getDx(front.getDisplayBounds(), xAlign, dx),
				Stack.getDy(front.getDisplayBounds(), yAlign, dy));
		double newLeft = Math.min(backBounds.getMinX(), frontBounds.getMinX());
		// TODO: why is this not getMaxX() for both?
		double newRight = Math.max(backBounds.getMinX() + backBounds.getWidth(), frontBounds.getMinX() + frontBounds.getWidth());
		double newTop = Math.min(backBounds.getMinY(), frontBounds.getMinY());
		double newBottom = Math.max(backBounds.getMaxY(), frontBounds.getMaxY());
		this.newBounds = new BoundingBox(newLeft, newTop, newRight - newLeft, newBottom - newTop);
		this.backTransform = Transform.translate(backBounds.getMinX() - newLeft,  backBounds.getMinY() - newTop);
		this.frontTransform = Transform.translate(frontBounds.getMinX() - newLeft, frontBounds.getMinY() - newTop);
	}
	
	public Stack(Image front, Image back) {
		this(front, back, XAlign.CENTER, YAlign.CENTER, 0.0, 0.0);
	}
	
	public Stack(Image front, Image back, Align align) {
		this(front, back, align.xAlign, align.yAlign, 0.0, 0.0);
	}
	
	public Shape getBounds() {
		return new Rectangle(newBounds.getMinX(), newBounds.getMinY(), newBounds.getWidth(), newBounds.getHeight());
	}
	
	private Node newNode() {
		Node backView = new Group(back.buildImage());
		backView.getTransforms().add(backTransform);
		Node frontView = new Group(front.buildImage());
		frontView.getTransforms().add(frontTransform);
		Pane pane = new Pane();
		pane.getChildren().add(new Group(backView, frontView));
		pane.setPrefWidth(newBounds.getWidth());
		pane.setPrefHeight(newBounds.getHeight());
		return pane;
	}
	
	public Node buildImage() {
		if (Platform.isFxApplicationThread()) {
			return newNode();
		} else {
			Task<Node> theNode = new Task<Node>() {
    			protected Node call() {
    				return Stack.this.newNode();
    			}
    		};
    		try {
    			Platform.runLater(theNode);
    			return theNode.get();
    		} catch (Exception e) {
    			throw new RuntimeException(e);
    		}
		}
	}

	private static BoundingBox translateRect(BoundingBox rect, double dx, double dy) {
		return new BoundingBox(rect.getMinX() + dx, rect.getMinY(), rect.getWidth(), rect.getHeight());
	}
	
	private static double getDx(BoundingBox bds, XAlign xAlign, double xOffset) {
		double amtToAdd;
		if (xAlign == XAlign.LEFT) {
			amtToAdd = 0;
		} else if (xAlign == XAlign.CENTER) {
			amtToAdd = -0.5 * bds.getWidth();
		} else { // (xAlign == XAlign.RIGHT)
			amtToAdd = -1.0 * bds.getMaxX();
		}
		return xOffset + amtToAdd;
    }
	
	private static double getDy(BoundingBox bds, YAlign yAlign, double yOffset) {
		double amtToAdd;
		if (yAlign == YAlign.TOP) {
			amtToAdd = 0;
		} else if (yAlign == YAlign.CENTER) {
			amtToAdd = -0.5 * bds.getHeight();
		} else { // (yAlign == YAlign.Bottom)
			amtToAdd = -1.0 * bds.getMaxY();
		}
		return yOffset + amtToAdd;
	}
	
	
}
